Implementing the Stopwatch

b0r1ngx opened the pull request:
Hello, Arkadii, how i said previously i want to add a stopwatch feature.

No matter how easy the task is, I decided to study different types of implementation, and think found one good idea (source: https://akjaw.com/kotlin-coroutine-flow-stopwatch-part1). I'm partly implemented same idea in core.

When i start implement and integrate logic in-to this project im get _stuck & overwhelmed_ (also with this mark, I'm created TODO's in code with some ideas), with thing of convenient and `correct` creating and usage the Stopwatch in a project.

Now the feature is not fully-integrated in the Game (it's not work when GameState goes to Failed or Win)

I will be appreciate to receive feedback from you, since you are the creator of the project and also have more experience in architectural design

b0r1ngx committed:
Implemented core-logic of Stopwatch and using it in project:
1) Added dependencies of kotlinx:
1.1) -datetime (to get time)
1.2) -coroutines-core (to count time with coroutines)

2) Implemented core logic of Stopwatch (wanted to make it like module - but make it later, because not so easy, for me, to adopt Stopwatch in project)

3) Using Stopwatch in project (now only start, when user pop first tile, and stop (reload) when user clicks `RestartButton`). To use Stopwatch in project:
3.1) include StopwatchViewModel in GameComponent
3.2) using collectAsState on the Flow of current tick of Stopwatch

Current bugs:
4) Stopwatch not stops even if user goes to GameStatus
4.1) WIN
4.2) FAILED

b0r1ngx committed:
delete unused feature of formatter in the Stopwatch

b0r1ngx committed:
not use wildcard imports

b0r1ngx committed:
1) Improved contentDescriptions of Counter's (not tested)
2) added description to TODO

b0r1ngx commented:
For now I'm want to write some tests for the Stopwatch.

b0r1ngx committed:
Refactoring & trying to provide right stopwatch state for PreviewGameComponent

b0r1ngx committed:
using coroutineContext.isActive instead of NonCancellable.isActive - for prevent failing build

arkivanov commented:
Thanks for working on this! I suggest to not over-engineer the timer. We can simply do the following.

1. Add a new property to `GameState` - e.g. `GameState.timer: Int` with the default value `0`.
2. Add a new Intent in GameStore - `Intent.TickTimer`.
3. Handle the new Intent - increment the `timer` value if its less than 999, or do nothing otherwise.
4. Add a new dependency: `com.arkivanov.essenty:lifecycle-coroutines:2.0.0-alpha02`.
5. In `DefaultGameComponent` - create a `private` `scope` using `com.arkivanov.essenty.lifecycle.coroutines.coroutineScope` function.
6. In DefaultGameComponent - just start/stop a coroutine job when `GameState.gameStatus` switches to/from STARTED. Inside that job - run an infinite loop with `delay(1.seconds)` and send `Intent.TickTimer` to `GameStore`.

Here is a rough example of how we can start/stop the job. It's from the top of my head, haven't tested it.

```kotlin
internal class DefaultGameComponent(...) {
    private val scope = coroutineScope()

    init {
        scope.launch {
            store.stateFlow
                .map { it.gameStatus == GameStatus.STARTED }
                .distinctUntilChanged()
                .collectLatest { isStarted ->
                    if (isStarted) {
                        while (true) {
                            delay(1.seconds)
                            store.accept(Intent.TickTimer)
                        }
                    }
                }
        }
    }
}
```

Optionally, we can cover this logic with integration tests.

1. Pass the main `CoroutineContext` in `DefaultGameComponent` class and use it in `coroutineScope`.
2. Create `DefaultGameComponentTest` class in `commonTest` source set.
3. Create `ComponentContext` using `DefaultComponentContext`.
4. Create `StoreFactory` using `DefaultStoreFactory`.
5. Use `StandardTestDispatcher` and pass it as `ComponentContext` to `DefaultGameComponent`.
6. Pass `TestCoroutineScheduler` to `StandardTestDispatcher` to control timings - you can advance the `TestCoroutineScheduler` timer.

Let me know if you have any questions!

b0r1ngx commented:
Glad to see you with information,

it seems that previously in project flow (and coroutines) isn't used at all, so using store.stateFlow doesn't allowed. If we doesn't import coroutines, how i already made before you write so now we can use flows, but did i must use some extension functions, to collect here State as Flow?

I tried to implement other logic inside scope, instead of that example that you give (doesn't think that it is right, or an idea inside is right, but i wanna just run a project with that):
```Kotlin
    init {
        scope.launch {
            state
                .map { it.gameStatus == GameStatus.STARTED }
                .let { isStarted ->
                    if (isStarted.value)
                        while (true) {
                            delay(1.seconds)
                            store.accept(Intent.TickTimer)
                        }
                }
        }
    }
```

Its compiled & builded, but at start of the app starts run I get this error:
[MVIKotlin]: Main thread id is undefined, main thread assert is disabled
Exception in thread "main" java.lang.IllegalStateException: Module with the Main dispatcher is missing. Add dependency providing the Main dispatcher, e.g. 'kotlinx-coroutines-android' and ensure it has the same version as 'kotlinx-coroutines-core'

tried to fix it with (at build.gradle.kts, composeApp module):
```Kotlin
        jvmMain.dependencies {
            implementation(compose.desktop.currentOs)
            implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.8.0")
//            implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core-macosArm64:1.8.0")
        }
```

Seems like the error is raising because of usage: com.arkivanov.essenty:lifecycle-coroutines:2.0.0-alpha02

I'll also try to find information at your Essenty-repo, but doesn't understand how to fix it

b0r1ngx commented:
Ahh.. lets I just commit, to you see more things, that i done by your guide

b0r1ngx committed:
1) Re-implement logic of stopwatch (timer)

2) Now the version is crashed with error: [MVIKotlin]: Main thread id is undefined, main thread assert is disabled
                                          Exception in thread "main" java.lang.IllegalStateException: Module with the Main dispatcher is missing. Add dependency providing the Main dispatcher, e.g. 'kotlinx-coroutines-android' and ensure it has the same version as 'kotlinx-coroutines-core'

arkivanov commented:
Oh, right. To fix the crash you need to add to jvm.main the following dependency: https://github.com/Kotlin/kotlinx.coroutines/tree/master/ui/kotlinx-coroutines-swing.
And please ignore the "Main thread id is undefined" message, it's just a warning.

b0r1ngx commented:
> Oh, right. To fix the crash you need to add to jvm.main the following dependency: https://github.com/Kotlin/kotlinx.coroutines/tree/master/ui/kotlinx-coroutines-swing. And please ignore the "Main thread id is undefined" message, it's just a warning.

Ah.. at the first my google-search-attempt I found this (https://github.com/JetBrains/compose-multiplatform/releases/tag/v1.1.1), but thinks, ehh, swing should it help us? - thinks no:c

Your advice is again helps! Its runs without crash, but yea, timer logic in scope is not works.

b0r1ngx committed:
fixed crash with using coroutines-swing import at jvmMain

b0r1ngx committed:
fixed `Reformat code` code transfer

arkivanov commented on composeApp/build.gradle.kts:
```diff
@@ -41,6 +41,9 @@ kotlin {
             implementation(libs.decompose.extensions.compose)
             implementation(libs.serialization.json)
+
+            implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.0")
```
I guess we don't need to add `kotlinx-coroutines-core` dependency? It should be already coming from Compose transitively.

b0r1ngx replied:
Hello, thanks for code-review, I'll check it now!

arkivanov commented on composeApp/build.gradle.kts:
```diff
@@ -41,6 +41,9 @@ kotlin {
             implementation(libs.serialization.json)
+
+            implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.0")
+            implementation("com.arkivanov.essenty:lifecycle-coroutines:2.0.0-alpha02")
```
Please add this dependency to `./gradle/libs.versions.toml`.

b0r1ngx replied:
Seems like @OptIn(ExperimentalComposeLibrary::class) using for implementation(compose.components.resources) (?) is `The opt-in annotation is redundant: no matching experimental API is used`, maybe we can remove annotation?

arkivanov replied:
Yeah, that's no longer experimental, we can remove it.

arkivanov commented on composeApp/build.gradle.kts:
```diff
@@ -49,6 +52,7 @@ kotlin {
 
         jvmMain.dependencies {
             implementation(compose.desktop.currentOs)
+            implementation("org.jetbrains.kotlinx:kotlinx-coroutines-swing:1.8.0")
```
Same here, please use `libs.versions.toml`.

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponent.kt:
```diff
@@ -24,6 +31,17 @@ internal class DefaultGameComponent(
+        scope.launch {
+            state
+                .map { it.gameStatus == GameStatus.STARTED }
+                .let { isStarted ->
```
This is incorrect. Please pay attention to the details in the code snippet [here](https://github.com/arkivanov/Minesweeper/pull/15#issuecomment-1986922426). This code missing `distinctUntilChanged()` and `collectLatest {...}`.

b0r1ngx replied:
Your code snippet seems awesome! I'm stuck with an idea in your code to gather stateFlow from a store

arkivanov replied:
Yes, I think you should also use `store.stateFlow` instead of `state`.

b0r1ngx replied:
How to convert store.state to Flow? I'm doesn't see any extension functions at Decompose library for Value type.

Should I change type of a state (start using Flow, instead of Value for GameComponent.state?) or create an extension function that create flow from GameComponent.state?

arkivanov replied:
Oh, I think we should also add `com.arkivanov.mvikotlin:mvikotlin-extensions-coroutines:4.0.0-alpha02` to `commonMain`. After that there should be `store.stateFlow` extension available.

b0r1ngx replied:
to be more, understandable, this is what I got, when write your snippet
<img width="873" alt="image" src="https://github.com/arkivanov/Minesweeper/assets/55315908/98e5815d-a16a-4b69-9b1e-0df34f825641">

b0r1ngx replied:
> Oh, I think we should also add `com.arkivanov.mvikotlin:mvikotlin-extensions-coroutines:4.0.0-alpha02` to `commonMain`. After that there should be `store.stateFlow` extension available.

yea, its what I talking about 😅, I'm not so aware of all your libraries that you made, let me try it!

b0r1ngx replied:
> Oh, I think we should also add `com.arkivanov.mvikotlin:mvikotlin-extensions-coroutines:4.0.0-alpha02` to `commonMain`. After that there should be `store.stateFlow` extension available.

Very thanks for opportunity of fast replies, I appreciate it.

b0r1ngx replied:
> This is incorrect. Please pay attention to the details in the code snippet [here](https://github.com/arkivanov/Minesweeper/pull/15#issuecomment-1986922426). This code missing `distinctUntilChanged()` and `collectLatest {...}`.

Awesome cache-hit in a first try of this code snippet:)

Also get a some not bad record in a first try:D
<img width="458" alt="image" src="https://github.com/arkivanov/Minesweeper/assets/55315908/f7b25cb9-b20f-495d-a49c-a4b2e30e4a4a">

I'd like to think about what we can do more for this project? Have you thought about it?

Maybe local leaderboard, I think it is still one of classic ways of realisation.

b0r1ngx replied:
Make a commit with update, where fix stopwatch! Now wants write some tests!

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/GameStore.kt:
```diff
@@ -211,3 +213,11 @@ private fun GameState.toggleFlagIntent(location: Location): GameState {
+        return this
+    }
+
+    return copy(timer = timer + 1)
```
Could be more Kotlin-idiomatic `if (timer < 999) copy(timer = timer + 1) else this`, and you can also use `=` instead of `return`.

b0r1ngx committed:
fixes given from code review

arkivanov commented on gradle/libs.versions.toml:
```diff
@@ -16,6 +16,8 @@ kotlin = "1.9.22"
 mvikotlin = "4.0.0-alpha02"
 serialization = "1.6.2"
+essenty-lifecycle-coroutines = "2.0.0-alpha02"
+coroutines-swing = "1.8.0"
```
Let's just name it `coroutines`, as this version could be applicable to all coroutines dependencies in the future.

arkivanov commented on gradle/libs.versions.toml:
```diff
@@ -16,6 +16,8 @@ kotlin = "1.9.22"
 decompose = "3.0.0-alpha05"
 mvikotlin = "4.0.0-alpha02"
 serialization = "1.6.2"
+essenty-lifecycle-coroutines = "2.0.0-alpha02"
```
Let's name it just `essenty`.

b0r1ngx committed:
1) Stopwatch is working - tested in game (Plans: write tests!)

*more fixes given from code review

b0r1ngx commented:
> Optionally, we can cover this logic with integration tests.
> 
> 1. Pass the main `CoroutineContext` in `DefaultGameComponent` class and use it in `coroutineScope`.
> 2. Create `DefaultGameComponentTest` class in `commonTest` source set.
> 3. Create `ComponentContext` using `DefaultComponentContext`.
> 4. Create `StoreFactory` using `DefaultStoreFactory`.
> 5. Use `StandardTestDispatcher` and pass it as `ComponentContext` to `DefaultGameComponent`.
> 6. Pass `TestCoroutineScheduler` to `StandardTestDispatcher` to control timings - you can advance the `TestCoroutineScheduler` timer.
> 
> Let me know if you have any questions!

Hello, I have a question about changes that we make for tests:

1. Passing `CoroutineContext` in `DefaultGameComponent`, should we also change signatures of `DefaultRootComponent` and function `DefaultRootComponent(componentContext: ComponentContext, storeFactory: StoreFactory): DefaultRootComponent`, to provide a `DefaultRootComponent` with `CoroutineContext` ?
1.1. Or we can use something like:
```Kotlin
gameComponentFactory = { ctx, settings ->
    DefaultGameComponent(componentContext = ctx, storeFactory = storeFactory, settings = settings, coroutineContext = SupervisorJob() + Dispatchers.Default)
},
```

2. If the first question is correctly resolved by me (check next commit), I'm done with your guide and now can start writing some integration tests.

b0r1ngx committed:
Preparations for later write tests for Stopwatch (timer)

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/root/DefaultRootComponent.kt:
```diff
@@ -60,7 +63,7 @@ internal fun DefaultRootComponent(componentContext: ComponentContext, storeFacto
         componentContext = componentContext,
         gameComponentFactory = { ctx, settings ->
-            DefaultGameComponent(componentContext = ctx, storeFactory = storeFactory, settings = settings)
+            DefaultGameComponent(componentContext = ctx, storeFactory = storeFactory, settings = settings, coroutineContext = SupervisorJob() + Dispatchers.Default)
```
We need to use `Dispatchers.Main.immediate` instead of `Default` to interact with the `GameStore`. I suggest to do the following.

1. Pass `Dispatchers.Main.immediate` to `DefaultGameComponent` here.
2. Add `SupervisorJob` inside `DefaultGameComponent` when calling `coroutineContext(...)`.
3. Rename `coroutineContext` argument to `mainCoroutineContext`.

We can also put every argument on separate line here for better readbility, as the line is pretty long otherwise.

arkivanov commented on composeApp/src/commonTest/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponentTest.kt:
```diff
@@ -0,0 +1,25 @@
+class DefaultGameComponentTest {
+    private val lifecycle = LifecycleRegistry()
+    private val componentContext = DefaultComponentContext(lifecycle = lifecycle)
+    private val storeFactory = DefaultStoreFactory()
```
No need for `componentContext` and `storeFactory` properties, they can be just inlined.

arkivanov commented on composeApp/src/commonTest/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponentTest.kt:
```diff
@@ -0,0 +1,25 @@
+            scheduler = coroutineScheduler
+        )
+    )
+
```
Please also add `fun before()` method annotated with `@BeforeTest` and call `lifecycle.resume()` from there.

arkivanov commented:
> Or we can use something like

Thanks for the update! The current approach looks good. `DefaultRootComponent` doesn't require `CoroutineContext`, so there is no need to pass it there. And no need to change `DefaultRootComponent` function for now.

b0r1ngx committed:
1) Changes given from code review
2) Wrote first test for stopwatch

b0r1ngx commented:
Hello, I'm trying to work with `TestCoroutineScheduler`, and understand how to shift it, shifting is happening good, but `DefaultGameComponent` itself looks do nothing, timer is just always on start (at zero) 

I'm tried to use:
1) `advanceTimeBy` for `coroutineScheduler`
2) `delay()` in `runTest { ... }`

Can you please give some suggestions, what we can do?

I'm tried to change gameStatus with next lines of code:
```Kotlin
var gameState = gameComponent.state.value
gameState = gameState.copy(gameStatus = GameStatus.STARTED)
```

Maybe I need to use something like you do at `GameStoreTest`, for change `GameState.gameStatus`, like `click` function?

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponent.kt:
```diff
@@ -21,10 +22,10 @@ internal class DefaultGameComponent(
 ) : GameComponent, ComponentContext by componentContext {
 
-    private val scope = coroutineScope(context = coroutineContext)
+    private val scope = coroutineScope(context = SupervisorJob() + mainCoroutineContext)
```
I'm not sure if it's related, but usually it's the opposite: `mainCoroutineContext + SupervisorJob()`.

b0r1ngx replied:
Hmm, okay, i make changes that you talking about, somewhere I saw how I'm declare it.
Is it just for best eye intercept to show that this context has also parent context, that coming first, before +?
Or it has some logic behind it, for coroutine work?

arkivanov replied:
I think it's the [docs](https://kotlinlang.org/docs/exception-handling.html#supervision). Also this way is [used](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:lifecycle/lifecycle-viewmodel/src/commonMain/kotlin/androidx/lifecycle/viewmodel/internal/CloseableCoroutineScope.kt;l=60) in android ViewModel.

arkivanov commented on composeApp/src/commonTest/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponentTest.kt:
```diff
@@ -2,24 +2,36 @@ package com.arkivanov.minesweeper.game
+
+    @Test
+    fun WHEN_created_THEN_stopwatch_on_START() {
+        val gameState = gameComponent.state.value
```
Let's extract a private property for easier access in every test:

```
private val state get() = gameComponent.state.value
```

b0r1ngx replied:
I like your idea because in my first vision I thought (tried) to hard-change `DefaultGameComponent` things, now I think that it is not how good tests must work, in this vision we don't show how different object interact with each other, even more we don't show how `DefaultGameComponent` itself does some things (like changing `gameStatus` or iterating `timer` (stopwatch)), how it is working in game, when we play it.

arkivanov commented:
> Can you please give some suggestions, what we can do?

It's hard to say without the code. I will try locally on your branch and update here. Thanks for working on this!

b0r1ngx committed:
Changes given from code review

arkivanov commented:
The following tests seems to work fine.

```kotlin
   @Test
    fun GIVEN_created_WHEN_cell_clicked_THEN_status_STARTED() {
        clickCellPrimary(x = 0, y = 0)

        assertEquals(GameStatus.STARTED, state.gameStatus)
    }

    @Test
    fun GIVEN_created_WHEN_cell_clicked_THEN_timer_shows_0() {
        clickCellPrimary(x = 0, y = 0)

        assertEquals(0, state.timer)
    }

    @Test
    fun GIVEN_game_started_WHEN_one_second_passed_THEN_timer_shows_1() {
        clickCellPrimary(x = 0, y = 0)

        coroutineScheduler.advanceTimeBy(1.seconds)
        coroutineScheduler.runCurrent()

        assertEquals(1, state.timer)
    }

    private fun clickCellPrimary(x: Int, y: Int) {
        gameComponent.onCellTouchedPrimary(x = x, y = y)
        gameComponent.onCellReleased(x = y, y = y)
    }
```

b0r1ngx commented:
It's an awesome possibility to start with, thanks for equip me with that 👨‍💻🎨

b0r1ngx committed:
1) Implemented some tests with project creator help
2) Deprecate using stopwatch in strings

arkivanov commented on composeApp/src/commonTest/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponentTest.kt:
```diff
@@ -31,10 +31,46 @@ class DefaultGameComponentTest {
+        coroutineScheduler.advanceTimeBy(10.seconds)
+        coroutineScheduler.runCurrent()
+
+        clickCellPrimary(x = 10, y = 10)
```
I think this cell is not necessarily a mine, as the test name says? It would be nice to test both cases: when we click on a mine (fail) and when we win the game. In this case, I think we can create the component with default settings.

We can change our `clickCellPrimary` function to the following:

```kotlin
    private fun clickCellPrimary(location: Location) {
        gameComponent.onCellTouchedPrimary(x = location.x, y = location.y)
        gameComponent.onCellReleased(x = location.x, y = location.y)
    }
```

So it can be called manually like `clickCellPrimary(0 by 0)`, and also we can find/iterate cells from the state and click.

For example, we can click on a mine as follows:

```kotlin
        val (mineLocation) = state.grid.entries.first { (_, cell) -> cell.value.isMine }
        clickCellPrimary(mineLocation)
```

And we can click on all non-mine cells as follows:

```kotlin
        state.grid.filterValues { it.status.isClosed && !it.value.isMine }.forEach { (location, _) ->
            clickCellPrimary(location)
        }
```

b0r1ngx committed:
1) Changes given from code review
2) shift timer to different values

arkivanov commented on composeApp/src/commonTest/kotlin/com/arkivanov/minesweeper/game/DefaultGameComponentTest.kt:
```diff
@@ -56,21 +56,39 @@ class DefaultGameComponentTest {
-        gameComponent.onCellTouchedPrimary(x = x, y = y)
-        gameComponent.onCellReleased(x = y, y = y)
+    @Test
+    fun GIVEN_game_started_WHEN_status_win_THEN_timer_STOPS() {
```
Sorry, last bit - it would be better to write "WHEN_win_THEN". Would it be possible to fix?

b0r1ngx replied:
sure, all is alright, im also agree with that, any tips appreciated! :)

b0r1ngx replied:
Thanks a lot for big help with implementing this feature  🍰

b0r1ngx committed:
status win -> win

arkivanov commented:
Thank you!
